src/libostree/ostree-sysroot-upgrader.c

```diff
@@ -541,6 +545,43 @@ ostree_sysroot_upgrader_pull_one_dir (OstreeSysrootUpgrader  *self,
         ostree_async_progress_finish (progress);
     }
 
+  /* Check to see if the commit marks the ref as EOL, redirecting to
+   * another. */
+  if (!ostree_repo_resolve_rev (repo, origin_refspec, FALSE,
+                                &new_revision, error))
+    return FALSE;
+
+  if (!ostree_repo_load_variant (repo,
+                                 OSTREE_OBJECT_TYPE_COMMIT,
+                                 new_revision,
+                                 &new_variant,
+                                 error))
+    return FALSE;
+
+  g_variant_get_child (new_variant, 0, "@a{sv}", &new_metadata);
+  rebase = g_variant_lookup_value (new_metadata, "ostree.endoflife-rebase", G_VARIANT_TYPE_STRING);
+  if (rebase)
+    {
+      const char *new_ref = g_variant_get_string (rebase, 0);
+
+      /* Pull the new ref */
+      if (self->origin_remote &&
+          (upgrader_flags & OSTREE_SYSROOT_UPGRADER_PULL_FLAGS_SYNTHETIC) == 0)
+        {
+          refs_to_fetch[0] = (char *) new_ref;
+          if (!ostree_repo_pull_one_dir (repo, self->origin_remote, dir_to_pull, refs_to_fetch,
+                                         flags, progress, cancellable, error))
+            return FALSE;
+        }
+
+        /* Use the new ref for the rest of the update process */
+        g_free (self->origin_ref);
+        self->origin_ref = g_strdup(new_ref);
+        g_free (origin_refspec);
+        origin_refspec = g_strconcat (self->origin_remote, ":", new_ref, NULL);
```

Comment:
Or possibly just do the whole block under `if (rebase && self->origin_remote)`?  I can't think of a use case for *local* refs being redirected.  But there's also an argument that we should be symmetric in our handling of local/remote refs - your change is definitely right from that perspective.